[Doc] Add documentation for TLS/BLS (#2990)
taichi-dev/taichi: docs/lang/articles/advanced/performance.md

    for i in range(8192):  # no decorator, use default settings
        ...
```

# Local Storage Optimizations

Taichi comes with a few optimizations that leverages the *fast memory* (e.g. CUDA shared memory, L1 cache) for performance optimization.

> The idea is straightforward: Where possible, Taichi substitues the accesses to the global memroy (slow) with that to the local one (fast),

Reviewer: line 73: 
`substitues` >>>` substitutes`
`memroy` >>> `memory` 
and a few typos detected by VS Code Spell Checker ;)
line 103: `exitting` >>> `exiting` 
line 120: `specificially` >>> `specifically`

Author: Thanks!

    for i in range(8192):  # no decorator, use default settings
        ...
```

> # Local Storage Optimizations

Reviewer 2: Maybe this can be changed to a smaller font. Same applies to the nested level. I find it a bit confusing to have two first level (#) title in one page.

> Taichi comes with a few optimizations that leverages the *fast memory* (e.g. CUDA

Reviewer 2: leverage

shared memory, L1 cache) for performance optimization. The idea is straightforward:

> Where possible, Taichi substitutes the access to the global memroy (slow) with that

Reviewer 2: Wherever possible.

> to the local one (fast), and writes the data in the local memory back to the global

Reviewer 2: ...in the local memory (e.g., CUDA shared memory) back...

> memory in the end. Such transformations preserves the semantics of the original

Reviewer 2: preserve

program (will be explained later).

## Thread Local Storage (TLS)

> TLS is mostly designed to optimize the parallel reduction. When Taichi identifies

Reviewer 2: optimize parallel reduction.

a global reduction pattern in a `@ti.kernel`, it automatically applies the TLS

> optimization to make the generated code match those that are commonly found in

Reviewer 2: it automatically applies the TLS optimizations during code generation similar to those found in common GPU reduction implementations.

s = ti.field(ti.f32, shape=())

@ti.kernel
def sum():
  for i in x:
    s[None] += x[i]

sum()
```

Internally, Taichi's parallel loop is implemented using
[Grid-Stride Loops](https://developer.nvidia.com/blog/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/).
What this means is that each physical CUDA thread could handle more than one item in `x`.
That is, the number of threads launched for `sum` can be fewer than the shape of `x`.

One optimization enabled by this strategy is to substitute the global memory access
with a *thread-local* one. Concretely, instead of directly and atomically adding
`x[i]` into the destination `s[None]`, which resides in the global memory, Taichi
preallocates a thread-local buffer upon entering the thread, accumulates
(*non-atomically*) the value of `x` into this buffer, then adds the result of the
buffer back to `s[None]` atomically before exiting the thread. Assuming each thread
handles `N` items in `x`, the number of atomic adds is reduced to one-N-th its
original size.

Additionally, the last atomic add to the global memory `s[None]` is optimized using
CUDA's warp-level intrinsics, further reducing the number of required atomic adds.

> Currently, Taichi supports TLS optimization for these reduction operators: add, sub,

Reviewer 2: ...for common reduction operators such as `add`, `sub`, `min`, and `max`.

## Block Local Storage (BLS)

Context: For a sparse field whose last layer is a `dense` SNode (i.e., its layer
hierarchy matches `ti.root.(sparse SNode)+.dense`), Taichi will assign one CUDA
thread block to each `dense` container (or `dense` block). BLS optimization works
specifically for such kinds of fields.

BLS aims to accelerate the stencil computation patterns by leveraging the CUDA
shared memory. This optimization starts with the users annotating the set of fields
they would like to cache via `ti.block_local`. Taichi then attempts to figure out
the accessing range w.r.t the `dense` block of these annotated fields at
*compile time*. If succeeded, Taichi generates code that first fetches all the
accessed data in range into a *block local* buffer (CUDA's shared memory), then
substitutes all the accesses to the corresponding slots into this buffer.

Here is an example illustrating the usage of BLS. `a` is a sparse field with a
block size of `4x4`.

```python
a = ti.field(ti.f32)
b = ti.field(ti.f32)
# `a` has a block size of 4x4
ti.root.pointer(ti.ij, 32).dense(ti.ij, 4).place(a)

@ti.kernel
def foo():
  # Taichi will cache `a` into the CUDA shared memory

>   ti.block_local(a)

Reviewer 2: highlight here?

  for i, j in a:
    print(a[i - 1, j], a[i, j + 2])
```

Each loop iteration accesses items with an offset `[-1, 0]` and `[0, 2]` to its

> coordinate, respectively. Therefore, for an entire block spanning from `[M, N]`

Reviewer 2: coordinates,

(inclusive) to `[M + 4, N + 4]` (exclusive), the accessed range w.r.t this block
is `[M - 1, M + 4) x [N, N + 6)` (derived from `[M + (-1), M + 4) x [N, N + 4 + 2)`).
The mapping between the global coordinates `i, j` and the local indices into the
buffer is shown below:

![](../static/assets/bls_indices_mapping.png)

From a user's perspective, you do not need to worry about these underlying details.
Taichi does all the inference and the global/block-local mapping automatically.

> That is, Taichi will pre-allocate a CUDA shared memory buffer of size `5x6`,

Reviewer 2: pre-allocate or preallocate? The one appeared before was preallocate.

pre-load `a`'s data into this buffer, and replace all the accesses to `a` (in the
global memory) with the buffer in the loop body. While this simple example does
not modify `a`, if a block-cached field does get written, Taichi would also generate
code that writes the buffer back to the global memory.

:::note

> BLS is not without its cost. Remember that BLS is designed for the stencil

Reviewer 3: Maybe `BLS has its cost` is more straightforward?

Author: ```suggestion
BLS does not come for free. Remember that BLS is designed for the stencil
```

> computation, where there is a large amount of overlapped access to the global

Reviewer 2: a large number of overlapped accesses...

> computation, where there are a large amount of overlapped accesses to the global

Reviewer 3: ```suggestion
computation, where there is a large amount of overlapped accesses to the global
```

Author: Note that it's *accesses* :joy:

Reviewer 3: I think the article `a` here is applied to `amount`. The set phrases `there is a large amount of` or `there are large amounts of ` seem to make more sense to me. But if the verb comes after `accesses`, I will vote for `are`, i.e., `A large amount of overlapped accesses are ...`. 
But anyway, I think either of the two choices here will not mislead the readers. 😂

Author: https://www.englishforums.com/English/ThereIsAreALargeNumberOf/bcqzgp/post.htm :joy:
